File: ObjcPodDemo/RixEngineObjcDemo/UI/Max/MaxNativeVC.m
#import "MaxNativeVC.h"
#import <AppLovinSDK/AppLovinSDK.h>
#import "AdConfig.h"

@interface MaxNativeVC () <MANativeAdDelegate>

@property (nonatomic, strong) UILabel *label;
@property (nonatomic, assign) BOOL isLoading;
@property (nonatomic, strong) UIView *adContainer;
@property (nonatomic, strong, nullable) MANativeAdLoader *adLoader;
@property (nonatomic, strong, nullable) MAAd *nativeAd;
@property (nonatomic, copy) NSString *TAG;

@end

@implementation MaxNativeVC

- (void)viewDidLoad {
    [super viewDidLoad];
    self.TAG = @"Max-native:";
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationItem.title = NSLocalizedString(@"max_native", @"");

    UIScrollView *scrollView = [[UIScrollView alloc] init];
    scrollView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:scrollView];
    scrollView.scrollEnabled = YES;

    UIStackView *contentView = [[UIStackView alloc] init];
    contentView.translatesAutoresizingMaskIntoConstraints = NO;
    [scrollView addSubview:contentView];
    contentView.axis = UILayoutConstraintAxisVertical;
    contentView.spacing = 20;

    UIButton *bnLoad = [self createButtonWithTitle:NSLocalizedString(@"load_ad", @"") action:@selector(buttonLoad)];
    [contentView addArrangedSubview:bnLoad];

    self.label = [self createLabel];
    [contentView addArrangedSubview:self.label];

    self.adContainer = [[UIStackView alloc] init];
    self.adContainer.translatesAutoresizingMaskIntoConstraints = NO;
    [contentView addArrangedSubview:self.adContainer];

    [NSLayoutConstraint activateConstraints:@[
        [scrollView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [scrollView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
        [scrollView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor],
        [scrollView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],

        [contentView.topAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.topAnchor constant:20],
        [contentView.leadingAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.leadingAnchor],
        [contentView.trailingAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.trailingAnchor],
        [contentView.bottomAnchor constraintEqualToAnchor:scrollView.contentLayoutGuide.bottomAnchor],
        [contentView.widthAnchor constraintEqualToAnchor:scrollView.frameLayoutGuide.widthAnchor],

        [bnLoad.heightAnchor constraintEqualToConstant:50],
        [self.label.heightAnchor constraintEqualToConstant:50],
        [self.adContainer.heightAnchor constraintEqualToConstant:500],
    ]];
}

- (void)buttonLoad {
    NSLog(@"bnLoad");
    if (self.isLoading) {
        return;
    }
    [self updateUI:YES message:NSLocalizedString(@"loading", @"")];
    [self loadAd];
}

- (void)updateUI:(BOOL)loading message:(NSString *)msg {
    self.isLoading = loading;
    self.label.text = msg;
}

- (void)loadAd {
    self.adLoader = [[MANativeAdLoader alloc] initWithAdUnitIdentifier:[AdConfig Max_Native_Ad_Id]];
    self.adLoader.nativeAdDelegate = self;
    [self.adLoader loadAd];
}

- (void)renderAdTemplatesUI:(MANativeAdView *)nativeAdView {
    [self clearSubView:self.adContainer];
    nativeAdView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.adContainer addSubview:nativeAdView];

    [NSLayoutConstraint activateConstraints:@[
        [nativeAdView.widthAnchor constraintEqualToAnchor:self.adContainer.widthAnchor],
        [nativeAdView.heightAnchor constraintEqualToConstant:500],
        [nativeAdView.topAnchor constraintEqualToAnchor:self.adContainer.topAnchor],
    ]];
}

#pragma mark - MANativeAdDelegate

- (void)didLoadNativeAd:(MANativeAdView *)nativeAdView forAd:(MAAd *)ad {
    NSLog(@"%@ didLoadNativeAd", self.TAG);
    [self updateUI:NO message:NSLocalizedString(@"load_success", @"")];

    if (self.nativeAd) {
        [self.adLoader destroyAd:self.nativeAd];
    }
    self.nativeAd = ad;

    if (nativeAdView) {
        [self renderAdTemplatesUI:nativeAdView];
    } else {
        NSLog(@"%@ no template ui", self.TAG);
    }
}

- (void)didFailToLoadNativeAdForAdUnitIdentifier:(NSString *)adUnitIdentifier withError:(MAError *)error {
    NSString *msg = [NSString stringWithFormat:@"%ld: %@", (long)error.code, error.description];
    NSLog(@"%@ didFailToLoadNativeAd: %@", self.TAG, msg);
    [self updateUI:NO message:[NSString stringWithFormat:NSLocalizedString(@"load_failed", @""), msg]];
}

- (void)didClickNativeAd:(MAAd *)ad {
    NSLog(@"%@ didClickNativeAd", self.TAG);
}

@end

File: SwiftPodDemo/AlxAdsDemo/UI/Max/MaxNativeVC.swift
//
//  MaxNativeVC.swift
//  AdsDemo
//
//  Created by liu weile on 2023/6/2.
//

import UIKit
import AppLovinSDK

class MaxNativeVC: BaseUIViewController {
    
    private let TAG = "Max-native:"

    private var label:UILabel!

    private var isLoading:Bool=false
    
    private var adContainer:UIView!

    private var adLoader:MANativeAdLoader?=nil
    private var nativeAd:MAAd?=nil

    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = .white
        navigationItem.title=NSLocalizedString("max_native", comment: "")
        
        let scrollView=UIScrollView()
        scrollView.translatesAutoresizingMaskIntoConstraints=false
        view.addSubview(scrollView)
        scrollView.isScrollEnabled=true
        
        let contentView=UIStackView()
        contentView.translatesAutoresizingMaskIntoConstraints=false
        scrollView.addSubview(contentView)
        contentView.axis = .vertical
        contentView.spacing = 20

        let bnLoad=createButton(title: NSLocalizedString("load_ad", comment: ""),  action: #selector(buttonLoad))
        contentView.addArrangedSubview(bnLoad)

        label=createLabel()
        contentView.addArrangedSubview(label)
        
        adContainer=UIStackView()
        adContainer.translatesAutoresizingMaskIntoConstraints=false
        contentView.addArrangedSubview(adContainer)

        NSLayoutConstraint.activate([
            scrollView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            scrollView.leadingAnchor.constraint(equalTo:view.safeAreaLayoutGuide.leadingAnchor),
            scrollView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
            scrollView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
            
            contentView.topAnchor.constraint(equalTo: scrollView.contentLayoutGuide.topAnchor,constant: 20),
            contentView.bottomAnchor.constraint(equalTo: scrollView.contentLayoutGuide.bottomAnchor),
            contentView.widthAnchor.constraint(equalTo: scrollView.frameLayoutGuide.widthAnchor),
            
            bnLoad.widthAnchor.constraint(equalTo: contentView.widthAnchor),
            bnLoad.heightAnchor.constraint(equalToConstant: 50),
            
            label.widthAnchor.constraint(equalTo: contentView.widthAnchor),
            label.heightAnchor.constraint(equalToConstant: 50),
            
            adContainer.widthAnchor.constraint(equalTo: contentView.widthAnchor),
            adContainer.heightAnchor.constraint(equalToConstant: 500),
        ])
    }
    
    @objc func buttonLoad(){
        print("bnLoad")
        if isLoading {
            return
        }

        updateUI(true, NSLocalizedString("loading", comment: ""))

        loadAd()
    }
    
    func updateUI(_ loading:Bool,_ msg:String){
        self.isLoading=loading
        self.label.text=msg
    }
    
    func loadAd() {
        adLoader=MANativeAdLoader(adUnitIdentifier: AdConfig.Max_Native_Ad_Id)
        adLoader?.nativeAdDelegate=self
        adLoader?.loadAd()
    }
    
    func renderAdTemplatesUI(_ nativeAdView: MANativeAdView){
        clearSubView(adContainer)
        nativeAdView.translatesAutoresizingMaskIntoConstraints=false
        adContainer.addSubview(nativeAdView)
        
        NSLayoutConstraint.activate([
            nativeAdView.widthAnchor.constraint(equalTo: adContainer.widthAnchor),
            nativeAdView.heightAnchor.constraint(equalToConstant: 500),
            nativeAdView.topAnchor.constraint(equalTo: adContainer.topAnchor),
        ])
    }

}

extension MaxNativeVC:MANativeAdDelegate{
    
    func didLoadNativeAd(_ nativeAdView: MANativeAdView?, for ad: MAAd) {
        print("\(TAG) didLoadNativeAd")
        updateUI(false, NSLocalizedString("load_success", comment: ""))
        
        if let currentNativeAd=nativeAd{
            adLoader?.destroy(currentNativeAd)
        }
        nativeAd=ad
        
        
        if let nativeAdView=nativeAdView{
            renderAdTemplatesUI(nativeAdView)
        }else{
            print("\(TAG) no template ui")
        }
    }
    
    func didFailToLoadNativeAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError) {
        let msg="\(error.code): \(error.description)"
        print("\(TAG) didFailToLoadNativeAd: \(msg)")
        updateUI(false, String(format: NSLocalizedString("load_failed", comment: ""), msg))
    }
    
    func didClickNativeAd(_ ad: MAAd) {
        print("\(TAG) didClickNativeAd")
    }
    
    
}
